### ISSUE TITLE
`ps_facetedsearch`: Sorting by price in faceted search does not work

### ISSUE THREAD

**Issue by Hlavtox** [PrestaShop/PrestaShop#19810]
Sorting by price in category (with faceted search) does not work.

### How to reproduce it
1. Install 1.7.7 beta 2
2. Install Faceted search 3.6.0
3. For simplicity. Keep only one language. Keep only one currency. Keep only one country active.
4. Make a customer group ID 4.
5. Make 3 products.
-- 5A. Product A, Price 100, 21% tax included.
-- 5B. Product B, Price 150, 21% tax included.
-- 5C. Product C, Price 200, 21% tax included.
-- 5D Set specific price 90 (21% tax included) for customer group 4.
-- 5E. Add products to some category.
6. Setup faceted search for this category. rebuild all indexes.
7. Go to FO, open this category and sort from the cheapest.
8. Order will look like this.

![order](https://user-images.githubusercontent.com/6097524/94155144-e6d0cf00-fe7e-11ea-8247-121dd3e5c619.PNG)

#### Acceptance Test
Unskip this test : 
https://github.com/PrestaShop/PrestaShop/blob/d672ad4ccc467722d13746e25fffc95a86e5e3b3/tests/UI/campaigns/functional/FO/classic/08_menuAndNavigation/02_sortAndFilter/01_sortProducts.ts#L164-L173

### Cause
Faceted search indexes (or it should, currently not working) the regular price AND the lowest price after catalog discounts or specific prices. It creates a range MIN-MAX. 

But, when sorting products, it sorts them by the MIN value in the index, so, when a product has some specific prices, the order is completely wrong.

It needs to be sorted by the prices the customer sees on the page.

### Additional information
* PrestaShop version: 1.7.7 beta 2
* Faceted search version: 3.6.0
* PHP version: 7.2

**Comment by rdy4ever** [comment]
Hi @Hlavtox 
I've never had a problem with faceted search sorting by price in any of my stores. Also, I can't reproduce this issue in a new install. So the problem must be with your particular installation/shop. Could you try to switch to the default theme and see if the problem persists?

**Comment by colinegin** [comment]
Hello @Hlavtox ,

which version of faceted search module are you using ? This seems to be working well with 3.5.0 version, i've just tested it.

Thanks !

**Comment by Hlavtox** [comment]
@colinegin See #19849, I found the cause.

**Comment by colinegin** [comment]
Hello @Hlavtox, I guess we can close this issue if this is a duplicate of the other issue you've created ?

**Comment by Hlavtox** [comment]
@PierreRambaud Hi, should we keep CRON and sorting issue separated? Because while indexing thing is fixable relatively easily, sorting is not, in my opinion.

**Comment by PierreRambaud** [comment]
Yes, we need to split issues as must as possible. Easier to test, easier to fix (when it's easy :sweat_smile: )

**Comment by Hlavtox** [comment]
@PierreRambaud Updated the issue description. After #19849 is fixed, how this can be solved?

**Comment by khouloudbelguith** [comment]
Hi @Hlavtox,

Your issue seems to be similar to this [ticket](https://github.com/PrestaShop/PrestaShop/issues/15223#issuecomment-523980939), is it?

Thanks!

**Comment by sarahdib** [comment]
@Hlavtox 
could you check the [issue](https://github.com/PrestaShop/PrestaShop/issues/15223#issuecomment-523980939) mentionned by @khouloudbelguith ? 
Thank you :)

**Comment by Hlavtox** [comment]
@sarahdib @khouloudbelguith The issue has the same cause, but was not fixed. Actually, this issue was introduced by fixing the related issue.

**Comment by khouloudbelguith** [comment]
Hi @Hlavtox,

I did not manage to reproduce the issue with PS1.7.6.7 and ps_facetedseach v3.5.0.
I attached a screen record
https://drive.google.com/file/d/1ZG6hWkoTUandczit_7f1RCz7V52V2Sb4/view?usp=sharing

Thanks to check and feedback.

**Comment by khouloudbelguith** [comment]
Hi @Hlavtox,

Since we had no news from you for more than 20 days, I'll close this ticket. Feel free to open another one if you can give specific details.

Thanks!

**Comment by Hlavtox** [comment]
Hi @khouloudbelguith, can you reopen this? The bug is there and needs to be fixed. Try to reproduce it with this scenario and I am sure you will see it. :-)

@PierreRambaud @eternoendless Can you have a look at it guys?

**Scenario**
1. Use the latest Prestashop and Faceted search dev code, but it should work with the released versions.
2. Make customer group ID 4.
3. Make 3 products.
3A. Product A, Price 100.
3B. Product B, Price 150. Specific price 90 for customer group 4.
3C. Product C, Price 200.
4. Setup faceted search, rebuild index and prices.
5. Go to FO, sort by price. Product B should be second, but is first.

**Comment by khouloudbelguith** [comment]
Hi @Hlavtox,

I tried with the release stable ps_facetedsearch v3.6.0. and the dev branch.
I have the same result
When I try to sort by Price from High to low, Product B is the second.
I attached a screen record
https://drive.google.com/file/d/1koVnQAal2C0_kJSzurTkf3mA-ywPOdBu/view?usp=sharing
Thanks to check and feedback.

**Comment by Hlavtox** [comment]
@khouloudbelguith I am not alone. 

@ioweb-gr has the same issue https://github.com/PrestaShop/PrestaShop/issues/21137

**Comment by ioweb-gr** [comment]
@Hlavtox I think the use cases are slightly different though that's why I opened other issue. I did provide detailed instructions on how to replicate my use case though and it's definitely a bug in the module. 

Also while debugging that, I noticed that during the upgrade process my table hadn't updated the table columns from INT(11) to DECIMAL(11) and I had to manually force the update on them. 

It seems that if you delete the table, the module won't recreate it. I would expect it to since it's an index table.

**Comment by Hlavtox** [comment]
@khouloudbelguith @PierreRambaud @ioweb-gr Just reproduced it on a clean PS install.

1. Install 1.7.7 beta 2
2. Install Faceted search 3.6.0
3. For simplicity. Keep only one language. Keep only one currency. Keep only one country active.
4. Make a customer group ID 4.
5. Make 3 products.
-- 5A. Product A, Price 100, 21% tax included.
-- 5B. Product B, Price 150, 21% tax included.
-- 5C. Product C, Price 200, 21% tax included.
-- 5D Set specific price 90 (21% tax included) for customer group 4.
-- 5E. Add products to some category.
6. Setup faceted search for this category. rebuild all indexes.
7. Go to FO, open this category and sort from the cheapest.
8. Order will look like this.

![order](https://user-images.githubusercontent.com/6097524/94155144-e6d0cf00-fe7e-11ea-8247-121dd3e5c619.PNG)

![db](https://user-images.githubusercontent.com/6097524/94155184-f3edbe00-fe7e-11ea-9092-88aa38b3e297.PNG)

**Comment by ioweb-gr** [comment]
Exactly this table needs to hold a column with the group id too to sort properly

**Comment by Hlavtox** [comment]
@ioweb-gr That will not help. It needs to resort it by the real displayed price. You say customer group, this all can alter the price you see.

- Country
- Customer group
- Currency
- Specific price for you only
- Specific prices can be limited by dates
- Catalog rules can mess with it

etc. etc. not easy bro 🤯 😄

**Comment by ioweb-gr** [comment]
Hehe yeah I wasn't looking for the rest use cases

**Comment by khouloudbelguith** [comment]
Hi @Hlavtox,

I followed these steps but it is ok in my case:
```
1. Install 1.7.7 beta 2
2. Install Faceted search 3.6.0
3. For simplicity. Keep only one language. Keep only one currency. Keep only one country active.
4. Make a customer group ID 4.
5. Make 3 products.
-- 5A. Product A, Price 100, 21% tax included.
-- 5B. Product B, Price 150, 21% tax included.
-- 5C. Product C, Price 200, 21% tax included.
-- 5D Set specific price 90 (21% tax included) for customer group 4.
-- 5E. Add products to some category.
6. Setup faceted search for this category. rebuild all indexes.
7. Go to FO, open this category and sort from the cheapest.
```

I attached a screen record
https://drive.google.com/file/d/1aA-YN9DI_ee8gwIM71twyzOcXR_yFXgw/view?usp=sharing

Thanks to check and feedback.

**Comment by Hlavtox** [comment]
@khouloudbelguith Nono don't do it by catalog rules. Open the product B and add a specific price on Pricing tab. Then rebuild index and go check the category, it's gonna do it... (I hope :-D)

![specific](https://user-images.githubusercontent.com/6097524/94487546-778e0e80-01e1-11eb-853e-076321506447.PNG)

**Comment by khouloudbelguith** [comment]
Hi @Hlavtox,

Thanks for your feedback.
I manage to reproduce the issue with PS1.7.6.8 and PS1770beta2 & ps_facetesearch v3.6.0. **But it is random**.
I attached a screen record
https://drive.google.com/file/d/1JhCSn4ORPuI-UUP--ct7GNm2bexMyhEj/view?usp=sharing

Record about the random issue
https://drive.google.com/file/d/1GfYOgUk0Z9ibtNgyapSxlsfdjuQC_Vbc/view?usp=sharing

Ping @PierreRambaud 
I’ll add this to the debug roadmap so that it’s fixed. If you have already fixed it on your end or if you think you can do it, please do send us a pull request!

Thanks!

**Comment by Hlavtox** [comment]
@khouloudbelguith It's random because you added the specific price to all 3 products.
If you keep it only on product B, the order will be like on my screenshot.
But you discovered another bug, if the products have the same price, they are sorted randomly.

I opened this in the past - https://github.com/PrestaShop/PrestaShop/issues/18674, It was closed, but clearly it's happening even on 1.7.7...

@PierreRambaud Check it when you have time, please... 🙏

**Comment by khouloudbelguith** [comment]
Hi!

I have a similar issue following these steps:
 

1. Fresh install PS1769, 177x, develop (1780)
2. ps_facetedsearch module is installed
3. Go to FO > Home category page 
4. Sort by price from high to low
5. See error 

![image](https://user-images.githubusercontent.com/16067358/111133092-69a72c80-857a-11eb-9360-52399f1c7fee.png)

I checked with ps_facetedsearch v3.6.0, v3.7.0, v3.7.1

Thank you!

**Comment by pab23** [comment]
Hi,

I still having this issue at Prestashop 1.7.7.2 and ps_facetedsearch 3.7.1

![image](https://user-images.githubusercontent.com/33161665/119799656-92359600-bedc-11eb-86d5-ef0e05d449cb.png)

Ordered by price low to high and not ordering poperly. 

This are the first four products 

![image](https://user-images.githubusercontent.com/33161665/119799930-d88af500-bedc-11eb-940f-e10e76d82de7.png)

Shouldn't the module sort with the current price for the specific customer?

Thanks!

**Comment by ioweb-gr** [comment]
So what's the status on this? Any roadmap or Eta or is the issue still unconfirmed by Prestashop?

**Comment by PierreRambaud** [comment]
> So what's the status on this? Any roadmap or Eta or is the issue still unconfirmed by Prestashop?

It's an open source project, feel free to contribute and fix the bug if you think it's mandatory for you :muscle:

**Comment by ioweb-gr** [comment]
My customer feels it's important and I agree it's really bad compared to other platforms to have such a bug. It's the first time I encounter such an issue which didn't exist in 1.6.  Unfortunately I lack the technical expertise in presta to handle it so I asked here if there was any update. There's no need for being aggressive in your response. An "unfortunately it's not that easy would suffice"

On our side we contacted many Prestashop developers for sponsoring a fix or replacing the module but unfortunately didn't find any possible solution that way either. Hence, why I asked for any possible news here.

**Comment by matks** [comment]
> There's no need for being aggressive in your response

Errr 😅  @PierreRambaud was not meant to be aggressive I believe. He just wanted to highlight that since this is an open source project, the quickest way to get something to be done is to push it yourself (which is a great strength when compared to a SAAS solution where you cannot modify the behavior yourself).

**Comment by PierreRambaud** [comment]
> My customer feels it's important and I agree it's really bad compared to other platforms to have such a bug. It's the first time I encounter such an issue which didn't exist in 1.6. Unfortunately I lack the technical expertise in presta to handle it so I asked here if there was any update. There's no need for being aggressive in your response. An "unfortunately it's not that easy would suffice"
> 
> On our side we contacted many Prestashop developers for sponsoring a fix or replacing the module but unfortunately didn't find any possible solution that way either. Hence, why I asked for any possible news here.

I'm sorry if I sounded aggressive, @matks is right, we can't fix everything alone, and a lot of bugs come from the previous versions (even in the 1.6).

Unfortunately, we don't plan to fix it in a short time :(

**Comment by jmirasolswg** [comment]
Hi everyone, I'm pretty new to Prestashop and I found this article. I have the same problem with a project I'm working on -- the sorting process behaves weird -- especially if there are products with discounts. I have been following this thread for quite some time now. I tried to experiment with the Faceted Search module and attempted to create a new sort order for discounts but as I found out while fiddling with the module:

- The sort orders must correspond to the table fields from the product tables. Since discounts are from specific_prices and there is not discounted value field in the tables (there is "reduction" in specific_prices by it does not tell the discounted value: you still have to subtract the price to it), the calculation only happens when the products are loaded in the product list, not during the sorting.

Still figuring it out how to possibly achieve it. 
I understand @PierreRambaud on his reply. If someone gets to solve it with the most efficient solution that would be great :) 

I'm working on PS 1.7.7.3.

**Comment by ioweb-gr** [comment]
I've tried contacting some of the more experienced development companies in the Prestashop ecosystem to see if anyone is willing to tackle this however I'm not sure if we could cover the budget needed for it

Hopefully finding someone who can solve it would be a good start though

@PierreRambaud no worries, I understand written words can be misinterpreted so I was wrong as well not trying to clarify your intentions. Anyways we all share the common goal of fixing this

**Comment by eternoendless** [comment]
We need some sort of crowdfunding service for this kind of stuff, where many people want something which is hard to do and can't individually finance it. But if they could agree on a communal fund, they could raise enough money to pay some high-end consultant to do it for all of them.

**Comment by Hlavtox** [comment]
I would be in, there is no way a successful client would choose Prestashop as their platform, if it cannot sort products by price. 👎

**Comment by roncarino** [comment]
Inside channel slack it has been brought to my attention this PR https://github.com/PrestaShop/PrestaShop/pull/18179 which seems to improve a lot the sorting experience in listing. The only problem left should be taking care also about discounted products (which does not for now).

**This has been merged already in 1.7.8**. Not entirely sure if this might be a great point to start to complete this issue too. And not too sure why this PR merged has not been mentioned here before. Maybe because it would be too "expensive" for faceted search to use the same logic for indexing? 🤔

**Comment by zulien** [comment]
I just checked this problem little deeper and here is partial solution. Not the best for performance but do the job.

Edit file:
/modules/ps_facetedsearch/src/Product/SearchProvider.php

Add on line 37.:
use Product;

And replace function:

    public function runQuery(
        ProductSearchContext $context,
        ProductSearchQuery $query
    ) {
        $result = new ProductSearchResult();
        // extract the filter array from the Search query
        $facetedSearchFilters = $this->filtersConverter->createFacetedSearchFiltersFromQuery($query);

        $context = $this->module->getContext();
        $facetedSearch = $this->searchFactory->build($context);
        // init the search with the initial population associated with the current filters
        $facetedSearch->initSearch($facetedSearchFilters);

        $orderBy = $query->getSortOrder()->toLegacyOrderBy(false);
        $orderWay = $query->getSortOrder()->toLegacyOrderWay();

        $filterProductSearch = new Filters\Products($facetedSearch);

        // get the product associated with the current filter
        $productsAndCount = $filterProductSearch->getProductByFilters(
            $query->getResultsPerPage(),
            $query->getPage(),
            $orderBy,
            $orderWay,
            $facetedSearchFilters
        );


for:

    public function runQuery(
        ProductSearchContext $context,
        ProductSearchQuery $query
    ) {
        $result = new ProductSearchResult();
        // extract the filter array from the Search query
        $facetedSearchFilters = $this->filtersConverter->createFacetedSearchFiltersFromQuery($query);

        $context = $this->module->getContext();
        $facetedSearch = $this->searchFactory->build($context);
        // init the search with the initial population associated with the current filters
        $facetedSearch->initSearch($facetedSearchFilters);

        $orderBy = $query->getSortOrder()->toLegacyOrderBy(false);
        $orderWay = $query->getSortOrder()->toLegacyOrderWay();

        $filterProductSearch = new Filters\Products($facetedSearch);

        // get the product associated with the current filter
        $productsAndCount = $filterProductSearch->getProductByFilters(
            $query->getResultsPerPage(),
            $query->getPage(),
            $orderBy,
            $orderWay,
            $facetedSearchFilters
        );

        if ($orderBy == 'price') {
            foreach($productsAndCount['products'] as $i=>$v) {
                $productsAndCount['products'][$i]['price'] = Product::getPriceStatic($v['id_product'], 0);
            }
            if ($orderWay == 'asc') {
                usort($productsAndCount['products'], function ($a, $b) { return ($a['price'] <=> $b['price']); });
            } else {
                usort($productsAndCount['products'], function ($a, $b) { return -($a['price'] <=> $b['price']); });
            }
        }

**Comment by Cyo90** [comment]
I was able to fix this again with the same method used by @zulien but for newer versions of Faceted Search (v3.12.1 currently).

File to edit:
/modules/ps_facetedsearch/src/Filters/Products.php

Search for (line 90):

`$matchingProductList = $this->searchAdapter->execute();`

And RIGHT AFTER that line, add this code:

    if ($orderBy == 'price') {
        foreach($matchingProductList as $i=>$v) {
            $matchingProductList[$i]['price'] = Product::getPriceStatic($v['id_product'], 0);
        }
        if ($orderWay == 'asc') {
            usort($matchingProductList, function ($a, $b) { return ($a['price'] <=> $b['price']); });
        } else {
            usort($matchingProductList, function ($a, $b) { return -($a['price'] <=> $b['price']); });
        }
    }

And done! Should work correctly now :)

**Comment by ioweb-gr** [comment]
@Cyo90 I tried your solution but unfortunately it won't work. You're just sorting the products after the filtering is done which means each page is sorted properly in it's own context. However if you start paginating you'll get different prices on page 1 and different on page 2. They won't be sorted accross all pages. 

Realistically speaking the only way to achieve this is to build a price index for each group and filter by the group id as well during sort

Alternatively you'd have to eliminate the pagination, sort the products, then cut down the results from the resulting array which would be a huge slowdown for large catalogs

**Comment by ioweb-gr** [comment]
I was wondering if here 

```
        if (isset($selectedFilters['price']) || $orderBy === 'price') {
            $this->searchAdapter->addSelectField('id_product');
            $this->searchAdapter->addSelectField('price');
            $this->searchAdapter->addSelectField('price_min');
            $this->searchAdapter->addSelectField('price_max');
        }
```

We can tamper with the joins and have a join on the ps_specific_price table to get the price filtered out from that instead of using price_min and price_max if a customer group is involved in the current customer context but unfortunately the searchAdapter doesn't provide any type of join. Also I don't see it possible to override the searchAdapter either for a temporary solution as it's not a class that can be overriden. 

It looks really limited in scope to actually do anything meaningful with it. 

Is there even a way to get the symfony container (and the entity manager) in that context as it's not a controller?

Looking at the core code it's all isolated and these are not even defined as services so I can't access the entity manager in any way that I can see :(

**Comment by ioweb-gr** [comment]
After a long time I still haven't found any developer who's capable / willing to tackle this. Any of the other people interested in this task had any luck with this? 

I believe it's a critical functionality to be able to sort by price according to the  prices you view in the current catalog page imho

**Comment by JMALDONADOM** [comment]
Prestashop 8.1.6 / ps_facetedsearch 3.16.1

File /modules/ps_facetedsearch/src/Filters/Products.php

Line 78…
if (isset($selectedFilters['price']) || $orderBy === 'price') {

change by…
if (isset($selectedFilters['price']) && $orderBy === 'price') {

work for me!

**Comment by lmartihohner** [comment]
Any solution?

**Comment by zulien** [comment]
Hello, this is still an issue. I just fixed it for PrestaShop v8.2.3 and ps_facetedsearch 4.0.1.

Edit this file:
/modules/ps_facetedsearch/src/Adapter/MySQL.php

Comment out line 390:

//$orderField = $this->getOrderDirection() === 'asc' ? 'price_min' : 'price_max';


This keeps the original price sorting instead of using the calculated price_min and price_max.

The problem is that the min/max calculation includes special prices, discounts, and similar rules. As a result, products may be sorted by discounted values even when those discounts are not applied for the current customer.

A better long-term solution might be to modify the layer_price_index table, for example by adding id_group, so the indexed prices match the customer group. For my case, disabling this sorting logic in the module was sufficient.

**Comment by boo-code** [comment]
The reason a group specific price never reaches the sort is in the shape of the index table, and I do not think it has been named in this thread.

Sorting by price reads the index, not the product price. `MySQL.php:477`:

```php
$orderField = $this->getOrderDirection() === 'asc' ? 'price_min' : 'price_max';
```

and `ps_layered_price_index` is keyed on:

```
PRIMARY KEY (id_product, id_currency, id_shop, id_country)
```

There is no group column.

The indexation does compute per group. It reads `id_group` from the specific prices (`ps_facetedsearch.php:457`) and passes it into the price computation at `:507` and `:542`. But the insert at `:599` writes only

```sql
INSERT INTO ps_layered_price_index (id_product, id_currency, id_shop, price_min, price_max, id_country)
VALUES ...
ON DUPLICATE KEY UPDATE id_product = id_product
```

Because the group is not part of the key, the rows computed for different groups collide on the same primary key, and `ON DUPLICATE KEY UPDATE id_product = id_product` keeps whichever arrived first and silently discards the rest. So the module does the group-aware work and then throws all but one result away.

That is why the reported setup behaves as it does: product C with a group 4 specific price of 90 is indexed once, under whichever group won the race, and the sort orders by that single value for every visitor.

It also explains why this looks inconsistent between shops. Which row survives depends on the order the groups come back in, so two shops with the same data can sort differently, and re-indexing can change the order on the same shop.

Fixing it properly means `id_group` in the table and in its primary key, in the insert, and in the query that reads it, plus a schema upgrade and a re-index. That is a real change to the module's storage rather than a patch, which is presumably why it has sat here, so I would rather check the shape is wanted before writing it. One question in particular: the index would grow by a factor of the number of groups carrying reductions, and I would want to know whether you would accept that, or prefer the sort to fall back to a live price computation when a group has specific prices.

For disclosure, I have an open pull request on this file, #1310, which touches the indexation loop around lines 396 to 476. The lines above are further down and do not overlap, but a fix here would want rebasing on it.
